src/environmentd/src/http/webhook.rs

```diff
@@ -0,0 +1,306 @@
+                "Invalid body type for Webhook source: {ty:?}"
+            ))?;
+        }
+    }
+
+    // Pack our headers into a row, if we're supposed to.
+    let mut headers_s = BTreeMap::new();
+    for (name, val) in headers.iter() {
+        if let Ok(val_s) = val.to_str() {
+            headers_s.insert(name.as_str(), val_s);
+        }
+    }
+    if header_ty.is_some() {
+        packer.push_dict(
+            headers_s
+                .iter()
+                .map(|(name, val)| (*name, Datum::String(*val))),
+        );
+    }
+
+    Ok(row)
+}
+
+/// Errors we can encounter when appending data to a Webhook Source.
+///
+/// Webhook sources are a bit special since they are handled by `environmentd` (all other sources
+/// are handled by `clusterd`) and data is "pushed" to them (all other source pull data). The
+/// errors also generally need to map to HTTP status codes that we can use to respond to a webhook
+/// request. As such, webhook errors don't cleanly map to any existing error type, hence the
+/// existence of this error type.
+#[derive(Error, Debug)]
+pub enum WebhookError {
+    #[error("no object was found at the path {}", .0.quoted())]
+    NotFound(String),
+    #[error("a resource was exhausted: {0}")]
+    ResourceExhausted(&'static str),
+    #[error("this feature is currently unsupported: {0}")]
+    Unsupported(&'static str),
+    #[error("failed to deserialize body as {ty:?}: {msg}")]
+    InvalidBody { ty: ScalarType, msg: String },
+    #[error("internal storage failure! {0:?}")]
+    InternalStorageError(StorageError),
+    #[error("internal adapter failure! {0:?}")]
+    InternalAdapterError(AdapterError),
+    #[error("internal failure! {0:?}")]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<StorageError> for WebhookError {
+    fn from(err: StorageError) -> Self {
+        match err {
+            // TODO(parkmycar): Maybe map this to a HTTP 410 Gone instead of 404?
+            StorageError::IdentifierMissing(id) => WebhookError::NotFound(id.to_string()),
+            StorageError::ResourceExhausted(rsc) => WebhookError::ResourceExhausted(rsc),
+            e => WebhookError::InternalStorageError(e),
+        }
+    }
+}
+
+impl From<AdapterError> for WebhookError {
+    fn from(err: AdapterError) -> Self {
+        match err {
+            AdapterError::Unsupported(feat) => WebhookError::Unsupported(feat),
+            e => WebhookError::InternalAdapterError(e),
+        }
+    }
+}
+
+impl IntoResponse for WebhookError {
+    fn into_response(self) -> axum::response::Response {
+        match self {
+            e @ WebhookError::NotFound(_) => (StatusCode::NOT_FOUND, e.to_string()).into_response(),
+            e @ WebhookError::ResourceExhausted(_) => {
+                (StatusCode::TOO_MANY_REQUESTS, e.to_string()).into_response()
+            }
+            e @ WebhookError::Unsupported(_) => {
+                (StatusCode::FORBIDDEN, e.to_string()).into_response()
+            }
```

Comment:
[405](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405) or [415](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/415) might be a more appropriate error code. I think 403 Forbidden indicates that the request would be possible if the client had the right privileges.

Reply:
Yeah I struggled with the right error code here. The reason I didn't use 405 was this part:

> The server **must** generate an Allow header field in a 405 status code response. The field must contain a list of methods that the target resource currently supports.

I guess actually we could generate an empty Allow header?

415 could also work, but doesn't seem exactly right either because of this statement:

> The format problem might be due to the request's indicated [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) or [Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding), or as a result of inspecting the data directly.

Reply:
> I guess actually we could generate an empty Allow header?

Probably not worth it.

What about just [400 - Bad Request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) then? Otherwise, I think 403 is probably good enough.